File: mobile_app/lib/features/operations/domain/operation_status.dart
import 'operation_type.dart';

enum OperationStatus {
  created,
  projectHeadApproval,
  customerApproval,
  waiting24Hours,
  completed,
  rejected,
  rolledBack;

  static OperationStatus fromJson(String value) {
    final v = value.trim().toUpperCase();
    return switch (v) {
      'CREATED'                => created,
      'PROJECT_HEAD_APPROVAL'  => projectHeadApproval,
      'CUSTOMER_APPROVAL'      => customerApproval,
      'WAITING_24_HOURS'       => waiting24Hours,
      'COMPLETED'              => completed,
      'REJECTED'               => rejected,
      'ROLLED_BACK'            => rolledBack,
      _                        => throw ArgumentError('Unknown OperationStatus: $value'),
    };
  }

  String toJson() => switch (this) {
        created             => 'CREATED',
        projectHeadApproval => 'PROJECT_HEAD_APPROVAL',
        customerApproval    => 'CUSTOMER_APPROVAL',
        waiting24Hours      => 'WAITING_24_HOURS',
        completed           => 'COMPLETED',
        rejected            => 'REJECTED',
        rolledBack          => 'ROLLED_BACK',
      };

  String get label => switch (this) {
        created             => 'Создана',
        projectHeadApproval => 'Ожидает руководителя',
        customerApproval    => 'Ожидает заказчика',
        waiting24Hours      => 'Период 24 часа',
        completed           => 'Завершена',
        rejected            => 'Отклонена',
        rolledBack          => 'Откат',
      };

  /// Default terminality for generic lifecycles (REJECTED is final until overridden per type).
  bool get isTerminalDefault => switch (this) {
        completed || rejected || rolledBack => true,
        _ => false,
      };

  /// Whether this status is terminal for [type]. Transfer treats [rejected] as intermediate.
  bool isTerminalForOperationType(OperationType type) => switch (type) {
        OperationType.transfer => switch (this) {
            completed || rolledBack => true,
            _ => false,
          },
        OperationType.income => switch (this) {
            completed => true,
            _ => false,
          },
        OperationType.report => isTerminalDefault,
      };
}

File: backend/app/Modules/Operations/Enums/OperationStatus.php
<?php

namespace App\Modules\Operations\Enums;

enum OperationStatus: string
{
    /** Newly created, awaiting first approval. */
    case CREATED                = 'CREATED';

    /** Pending approval by the project head (PROJECT_HEAD). */
    case PROJECT_HEAD_APPROVAL  = 'PROJECT_HEAD_APPROVAL';

    /** Pending approval by the customer (CUSTOMER). */
    case CUSTOMER_APPROVAL      = 'CUSTOMER_APPROVAL';

    /** Both parties approved; 24-hour rollback window is open. */
    case WAITING_24_HOURS       = 'WAITING_24_HOURS';

    /** Successfully completed; balances finalised. */
    case COMPLETED              = 'COMPLETED';

    /** Rejected at any approval stage; no balance change. */
    case REJECTED               = 'REJECTED';

    /** Rolled back during the 24-hour window; balances reversed. */
    case ROLLED_BACK            = 'ROLLED_BACK';

    /**
     * Default terminality for types that follow the generic lifecycle map
     * ({@see OperationTransitionService}) and as fallback for Income/Report until overridden.
     */
    public function isTerminal(): bool
    {
        return match ($this) {
            self::COMPLETED, self::REJECTED, self::ROLLED_BACK => true,
            default => false,
        };
    }

    /**
     * Terminality for a concrete operation type. Transfer treats REJECTED as non-terminal
     * (ТЗ-05.2: PROJECT_HEAD_APPROVAL → REJECTED → CREATED). Income/Report use {@see isTerminal()}
     * until their policies diverge.
     */
    public function isTerminalForOperationType(OperationType $type): bool
    {
        return match ($type) {
            OperationType::TRANSFER => match ($this) {
                self::COMPLETED, self::ROLLED_BACK => true,
                default => false,
            },
            OperationType::INCOME => match ($this) {
                self::COMPLETED => true,
                default => false,
            },
            OperationType::REPORT => $this->isTerminal(),
        };
    }
}
